roles/openafs_devel/library/openafs_build.py

```diff
@@ -0,0 +1,652 @@
+import shutil
+from multiprocessing import cpu_count
+from ansible.module_utils.basic import AnsibleModule
+
+
+class FileError(Exception):
+    pass
+
+def copy_tree(src, dst):
+    """Copy an entire directory tree.
+
+    Creates destination if needed. Clobbers any existing files/symlinks.
+
+    :arg src: directory to copy from. must already exist
+    :arg dst: directory to copy to. created if not already present
+    :returns: a list of files/symlinks created
+    """
+    outputs = []
+    if not os.path.isdir(src):
+        raise FileError("cannot copy tree '%s': not a directory" % src)
+    try:
+        names = os.listdir(src)
+    except os.error:
+        raise FileError("error listing files in '%s'" % src)
+    if not os.path.isdir(dst):
+        os.makedirs(dst)
+    for n in names:
+        src_name = os.path.join(src, n)
+        dst_name = os.path.join(dst, n)
+        if os.path.islink(src_name):
+            link_dest = os.readlink(src_name)
+            if os.path.exists(dst_name):
+                os.remove(dst_name)
+            os.symlink(link_dest, dst_name)
+            outputs.append(dst_name)
+        elif os.path.isdir(src_name):
+            outputs.extend(copy_tree(src_name, dst_name))
+        else:
+            shutil.copy2(src_name, dst_name)
+            outputs.append(dst_name)
+    return outputs
+
+def abspath(base, rel):
+    """ Get absolute path name relative to a base directory. """
+    prev = os.getcwd()
+    os.chdir(base)
+    path = os.path.abspath(rel)
+    os.chdir(prev)
+    return path
+
+def tail(s, n=256):
+    """ Get the last n chars of a string. """
+    if len(s) <= n:
+        return s
+    else:
+        return s[-n:]
+
+def run_command(name, command, cwd, module, logger, logdir, results):
+    """Run a command and log the stdout and stderr.
+
+    :arg command: command argument list
+    :arg cwd: current directory to run the command
+    :arg module: the ansible module object
+    :arg logger: the logger object
+    :arg logdir: where to place stdout and stderr logs
+    :arg results: the module results dictionary
+    """
+    logger.info('Running: cd %s; %s' % (cwd, ' '.join(command)))
+    rc, out, err = module.run_command(command, cwd=cwd)
+    logfile_out = os.path.join(logdir, '%s.out' % name)
+    with open(logfile_out, 'w') as f:
+        f.write(out)
+        results['changed'] = True
+        results['logfiles'].append(logfile_out)
+    logfile_err = os.path.join(logdir, '%s.err' % name)
+    with open(logfile_err, 'w') as f:
+        f.write(err)
+        results['changed'] = True
+        results['logfiles'].append(logfile_err)
+    if rc != 0:
+        logger.error('%s failed; rc=%d' % (name, rc))
+        module.fail_json(
+            msg='%s command failed. See log files %s and %s' % \
+                (name, logfile_out, logfile_err),
+            rc=rc,
+            stdout=tail(out),
+            stderr=tail(err),
+        )
+
+def _od2a(options, prefix=None):
```

Comment:
Maybe have these functions also stuffed into another library file? Might simplify the "main" program.  Though I would understand keeping it all in here as well..

Reply:
Again, that would require infrastructure to create and maintain separate python package, which would really complicate installation and maintenance. Maybe in a future release, but i'd rather not introduce that complexity yet.

Reply:
Understand.  Wasn't quite sure how much you wanted to do with the initial commit on these and what direction you wanted to go towards.

Reply:
well, if you look at the playbooks/build*.yaml, thats were the real action is, as that's some example on how these modules would be used in your playbooks.

Reply:
started looking at the build* playbooks